airbyte-ci/connectors/pipelines/pipelines/gradle.py

```diff
@@ -50,66 +53,107 @@ def build_include(self) -> List[str]:
+            # Mount the parts of the repo which interest us in /airbyte.
+            .with_exec(["mkdir", "/airbyte"])
+            .with_workdir("/airbyte")
+            .with_mounted_directory("/airbyte", self.context.get_repo_dir(".", include=include))
             .with_mounted_directory(str(self.context.connector.code_directory), await self.context.get_connector_dir())
             # Disable the Ryuk container because it needs privileged docker access that does not work:
             .with_env_variable("TESTCONTAINERS_RYUK_DISABLED", "true")
-            .with_(environments.mounted_connector_secrets(self.context, f"{self.context.connector.code_directory}/secrets"))
+            # Run gradle once to populate the container's local maven repository.
+            # This step is useful also to serve as a basic sanity check and to warm the gradle cache.
+            # This will download gradle itself, a bunch of poms and jars, compile the gradle plugins, configure tasks, etc.
+            .with_exec(self._get_gradle_command(":airbyte-cdk:java:airbyte-cdk:publishSnapshotIfNeeded"))
         )
-        if self.with_java_cdk_snapshot:
-            connector_under_test = connector_under_test.with_exec(["./gradlew", ":airbyte-cdk:java:airbyte-cdk:publishSnapshotIfNeeded"])
-        connector_under_test = connector_under_test.with_exec(self._get_gradle_command())
-
-        results = await self.get_step_result(connector_under_test)
 
-        await self._export_gradle_dependency_cache(connector_under_test)
-        return results
-
-    async def _export_gradle_dependency_cache(self, gradle_container: Container) -> Container:
-        """Export the Gradle writable dependency cache to the read-only dependency cache path.
-        The read-only dependency cache is persisted thanks to mounted cache volumes in environments.with_gradle().
-        You can read more about Shared readonly cache here: https://docs.gradle.org/current/userguide/dependency_resolution.html#sub:shared-readonly-cache
-        Args:
-            gradle_container (Container): The Gradle container.
-
-        Returns:
-            Container: The Gradle container, with the updated cache.
-        """
-        try:
-            cache_dirs = await gradle_container.directory(consts.GRADLE_CACHE_PATH).entries()
-        except QueryError:
-            cache_dirs = []
-        if "modules-2" in cache_dirs:
-            with_cache = gradle_container.with_exec(
-                [
-                    "rsync",
-                    "--archive",
-                    "--quiet",
-                    "--times",
-                    "--exclude",
-                    "*.lock",
-                    "--exclude",
-                    "gc.properties",
-                    f"{consts.GRADLE_CACHE_PATH}/modules-2/",
-                    f"{consts.GRADLE_READ_ONLY_DEPENDENCY_CACHE_PATH}/modules-2/",
-                ]
+        # From this point on, we add layers which are task-dependent.
```

Comment:
For task-dependent layers, what do you think about declaring them in the subclasses `_run`? 
I think it will help define a cleaner "base" container with gradle inside it. 
E.G we might need this class with a non-connector related context which does not have the `context.connector` attribute. The type hints on the `context` attribute show that a `PipelineContext` is expected in the constructor. If I'm not mistaken the `connector` attribute is not available in the superclass, only on the `ConnectorContext` subclass which inherits from it.

Reply:
Perhaps, or perhaps expose a class method to inject extra prelude steps. I don't have strong opinions in this regard.